cmd/object-handlers.go

```diff
@@ -302,46 +299,33 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
 		return
 	}
 
-	// Size of object.
-	size := objInfo.Size
-
-	pipeReader, pipeWriter := io.Pipe()
-	go func() {
-		startOffset := int64(0) // Read the whole file.
-		// Get the object.
-		gErr := objectAPI.GetObject(sourceBucket, sourceObject, startOffset, size, pipeWriter)
-		if gErr != nil {
-			errorIf(gErr, "Unable to read an object.")
-			pipeWriter.CloseWithError(gErr)
+	srcMetadata := objInfo.UserDefined
+	updateMetadata := isMetadataUpdate(srcMetadata, newMetadata)
+	if updateMetadata {
+		// Merge any new metadata.
+		mergeNewMetadata(srcMetadata, newMetadata)
+	} else {
+		// Source and destination objects cannot be same if metadata
+		// is not being updated.
+		if srcObject == dstObject && srcBucket == dstBucket {
+			writeErrorResponse(w, r, ErrInvalidCopyDest, r.URL.Path)
 			return
 		}
-		pipeWriter.Close() // Close.
-	}()
-
-	// Save other metadata if available.
-	metadata := objInfo.UserDefined
-
-	// Remove the etag from source metadata because if it was uploaded as a multipart object
-	// then its ETag will not be MD5sum of the object.
-	delete(metadata, "md5Sum")
-
-	sha256sum := ""
-
-	objectWLock := globalNSMutex.NewNSLock(bucket, object)
-	objectWLock.Lock()
-	defer objectWLock.Unlock()
+		// Hold write locks on destination if metadata is not to be updated
+		// on source but instead destination object is being created.
+		objectDWLock := globalNSMutex.NewNSLock(dstBucket, dstObject)
+		objectDWLock.Lock()
+		defer objectDWLock.Unlock()
+	}
 
-	// Create the object.
-	objInfo, err = objectAPI.PutObject(bucket, object, size, pipeReader, metadata, sha256sum)
+	// Copy source object to destination, if `updateMetadata` is set
+	// only metadata is updated. Source object and destination object
+	// should be same when metadata is being updated.
+	objInfo, err = objectAPI.CopyObject(srcBucket, srcObject, dstBucket, dstObject, srcMetadata, updateMetadata)
```

Comment:
`objInfo, err = objectAPI.CopyObject(srcBucket, srcObject, dstBucket, dstObject, mergedMetadata)`

will be a better API because, to decide whether to update just the metadata or not, the object layer can derive it by comparing src/dst